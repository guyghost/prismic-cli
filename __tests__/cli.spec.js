/**
 * comands
 *
 * prismic login [ --email | --password | --oauthaccesstoken ]
 * prismic quickstart [--folder | --template | --new]
 * prismic init [ --folder ]
 * prismic new [ --folder ]
 * prismic theme [ --theme-url | --folder | --conf | --template ]
 * prismic logout
 * prismic signup
 * prismic list
 * prismic --version
 * prismic sm --help
 * primsic sm --ls
 * prismic sm --setup [ --no-prismic | --library | --lib | --local-path ]
 * prismic sm --bootstrap
 * prismic sm --create-slice [ --template-path | --framework ]
 * prismic sm --add-storybook [ --framework ]
 * prismic sm --pull
 *
 * [UN-DOCUMENTED]
 * prismic sm --sm-endpoint
 * prismic sm --override
 * prismic sm --yes|| -y || --noconfirm
 * prismic --debug
 * prismic base || --base
 * prismic --users
 * prismic --noconfirm
 * prismic --domain
 * prismic --endpoint
 * prismic --help
 * prismic heroku
 * 
 * */
const npm = require('npm');
const path = require('path');
const fs = require('fs');
const { promisify } = require('util');
const { spawnSync } = require('child_process');

const TMP_DIR = path.resolve('__tmp__');
const PRISMIC_BIN = require.resolve('../bin/prismic');

/*
const util = require('util');
const exec = util.promisify(spawn); */

beforeAll(() => {
  const rmdir = promisify(fs.rmdir);
  const mkdir = promisify(fs.mkdir);
  const chmod = promisify(fs.chmod);
  const load = promisify(npm.load);
  const run = promisify(npm.run);

  return rmdir(TMP_DIR, { recursive: true })
    .then(() => mkdir(TMP_DIR))
    .then(() => load())
    .then(() => run('build'))
    .then(() => chmod(PRISMIC_BIN, 0o755));
});

describe('prismic --help', () => {
  test('it should write usage instructions to stdout', async () => {
    const res = spawnSync(PRISMIC_BIN, ['--help'], { encoding: 'utf8' /* ,stdio: 'inherit' */ });
    expect(res.stdout).toBeTruthy();
    expect(res.stdout).toMatchSnapshot();
    // expect(stderr).toBeNull(); lots of warnings :/
    // expect(res.stderr).toMatchSnapshot();
  });
});

describe('prismic --version', () => {
  it('should print out the current version on the cli', () => {
    const args = ['--version'];
    const res = spawnSync(PRISMIC_BIN, args, { encoding: 'utf8' });
    // eslint-disable-next-line global-require
    const { version } = require('../package.json');
    const str = res.stdout.replace(/\r|\n/g, '');
    expect(str).toBe(version);
    expect(res.stderr).toBeFalsy();
  });
});

describe('prismic list', () => {
  it('should list the avaible templates', () => {
    const { stdout } = spawnSync(PRISMIC_BIN, ['list'], { encoding: 'utf8' });
    expect(stdout.includes('NodeJS')).toBe(true);
    expect(stdout.includes('React')).toBe(true);
    expect(stdout.includes('Angular2')).toBe(true);
    expect(stdout.includes('Vue.js')).toBe(true);
  });
});


describe.skip('prismic login [ --email | --password | --oauthaccesstoken ]', () => {});

/* requires login */
describe.skip('prismic new', () => {
  const dir = path.join(TMP_DIR, 'test-new');
  const repoName = `qwerty-${Math.random().toString(36).replace(/[^a-z]+/g, '').substr(0, 5)}`;
  const args = ['new', '--domain', repoName, '--folder', dir, '--template', 'NodeJS'];

  it('should initialise a project from a template', () => {
    const res = spawnSync(PRISMIC_BIN, args, { encoding: 'utf8', shell: true });
    expect(res.stdout).toMatchSnapshot();
    expect(fs.existsSync(dir)).toBeTruthy();
  });
});

/* requires login */
describe.skip('prismic init', () => {
  const dir = path.join(TMP_DIR, 'init-test');
  const repoName = 'qwerty2223445566';
  const args = ['init', '--domain', repoName, '--folder', dir, '--template', 'NodeJS', '--noconfirm'];

  it('should initialise a project from a template', () => {
    const res = spawnSync(PRISMIC_BIN, args, { encoding: 'utf8', shell: true });
    expect(res.stdout).toMatchSnapshot();
    expect(fs.existsSync(dir)).toBeTruthy();
  });
});

/* requires login */
describe.skip('prismic quickstart [--folder | --template | --new]', () => {
  const dir = path.join(TMP_DIR, 'quickstart');
  const randomString = Math.random().toString(36).replace(/[^a-z]+/g, '').substr(0, 5);
  const repoName = `quickstart-test-${randomString}`;
  const args = ['quickstart', '--folder', dir, '--domain', repoName, '--template', 'NodeJS'];

  it('should initialize a project and repository', () => {
    const res = spawnSync(PRISMIC_BIN, args, { encoding: 'utf8' });
    expect(res.status).toBeFalsy();
    expect(res.stdout).toBeTruthy();
  });
});

/* requires login */
describe.skip('prismic theme [ --theme-url | --folder | --conf | --template ]', () => {
  it('should reate a rpoject using a theme', () => {
    const dir = path.join(TMP_DIR, 'theme-nuxt');
    const randomString = Math.random().toString(36).replace(/[^a-z]+/g, '').substr(0, 5);
    const repoName = `theme-test-${randomString}`;
    const args = [
      'theme',
      '--theme-url', 'https://github.com/prismicio/nuxtjs-blog.git',
      '--conf', 'nuxt.config.js',
      '--domain', repoName,
      '--folder', dir,
    ];

    const res = spawnSync(PRISMIC_BIN, args, { encoding: 'utf8' });
    expect(res.stdout).toBeTruthy();
    expect(fs.existsSync(dir)).toBe(true);
  })
});

describe.skip('prismic slicemachine', () => {
  it('prismic sm --help', () => {});
  it('prismic sm --ls', () => {});
  it('prismic sm --setup [ --no-prismic | --library | --lib | --local-path ]', () => {});
  it('prismic sm --bootstrap', () => {});
  it('prismic sm --create-slice [ --template-path | --framework ]', () => {});
  it('prismic sm --add-storybook [ --framework ]', () => {})
  it('prismic sm --pull', () => {});
});

describe.skip('prismic logout', () => {});
describe.skip('prismic signup', () => {});