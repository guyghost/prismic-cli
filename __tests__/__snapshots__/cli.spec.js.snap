// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prismic --help it should write usage instructions to stdout 1`] = `
"
Synopsis

  $ prismic <command> <domain> <options> 

Examples

  $ prismic init                          Create a project from an existing     
                                          prismic repository.                   
  $ prismic new                           Create a project with a new prismic   
                                          repository.                           
  $ prismic init foobar                   Create a project for the foobar       
                                          repository.                           
  $ prismic init foobar --folder          Create a NodeJS project, non-         
  ~/Desktop/myProject --template NodeJS   interactive.                          
  --noconfirm                                                                   
  $ prismic theme                         Create a project from a zip file with 
  https://github.com/prismicio/nodejs-    a new prismic repository.             
  sdk                                                                           
  $ prismic sm --setup                    Setup a nuxt.js project with          
                                          SliceMachine.                         

Command List

  quickstart          Create a project: initialize a node.js quickstart project 
                      with a new prismic repository.                            
  init                Initialize a project: initialize the code from a template 
                      for an existing prismic repository.                       
  new                 Create a project: initialize the code for a new prismic   
                      repository.                                               
  theme               Create a project: initialize project from a theme with a  
                      new prismic repository.                                   
  login               Login to an existing prismic.io account.                  
  logout              Logout from an existing prismic.io account.               
  signup              Create a new prismic.io account.                          
  list                List the available code templates.                        
  slicemachine | sm   Run slicemachine                                          

Options

  --email        The email of the account to use.                               
  --password     The password of the account to use.                            
  --new                                                                         
  --users                                                                       
  --folder       The folder to create the new project.                          
  --template     Project template to use (see the list command for available    
                 templates).                                                    
  --theme-url                                                                   
  --conf         Specify path to prismic configuration file. Used with \`theme\`  
                 command.                                                       
  --version                                                                     

"
`;

exports[`prismic init should initialise a project from a template 1`] = `
"Let's get to it!
"
`;

exports[`prismic new should initialise a project from a template 1`] = `
"Let's get to it!
Initialize local project
Running npm install...
Your project is ready, to proceed:

Go to the project folder : cd /Users/marc/Projects/prismic/prismic-cli/__tmp__/test-new
To have hot reload: npm install -g nodemon
Start your project: nodemon app.js
Point your browser to: http://localhost:3000
"
`;
